x/precisebank/types/fractional_balance_test.go

```diff
@@ -0,0 +1,140 @@
+package types_test
+
+import (
+	"testing"
+
+	sdkmath "cosmossdk.io/math"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/kava-labs/kava/app"
+	"github.com/kava-labs/kava/x/precisebank/types"
+	"github.com/stretchr/testify/require"
+)
+
+func TestNewFractionalBalance(t *testing.T) {
+	tests := []struct {
+		name        string
+		giveAddress string
+		giveAmount  sdkmath.Int
+	}{
+		{
+			"correctly sets fields",
+			"cosmos1qperwt9wrnkg5k9e5gzfgjppzpqur82k6c5a0n",
+			sdkmath.NewInt(100),
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			fb := types.NewFractionalBalance(tt.giveAddress, tt.giveAmount)
+
+			require.Equal(t, tt.giveAddress, fb.Address)
+			require.Equal(t, tt.giveAmount, fb.Amount)
+		})
+	}
+}
+
+func TestFractionalBalance_Validate(t *testing.T) {
+	app.SetSDKConfig()
+
+	tests := []struct {
+		name        string
+		giveAddress string
+		giveAmount  sdkmath.Int
+		wantErr     string
+	}{
+		{
+			"valid",
+			"kava1gpxd677pp8zr97xvy3pmgk70a9vcpagsakv0tx",
+			sdkmath.NewInt(100),
+			"",
+		},
+		{
+			"valid - 0 balance",
+			"kava1gpxd677pp8zr97xvy3pmgk70a9vcpagsakv0tx",
+			sdkmath.NewInt(0),
+			"",
+		},
+		{
+			"valid - max balance",
+			"kava1gpxd677pp8zr97xvy3pmgk70a9vcpagsakv0tx",
+			types.MAX_FRACTIONAL_AMOUNT,
+			"",
+		},
+		{
+			"invalid - uppercase letter",
+			"kava1gpxd677pP8zr97xvy3pmgk70a9vcpagsakv0tx",
+			sdkmath.NewInt(100),
+			"decoding bech32 failed: string not all lowercase or all uppercase",
+		},
+		{
+			"invalid - non-bech32 address",
+			"invalid",
+			sdkmath.NewInt(100),
+			"decoding bech32 failed: invalid bech32 string length 7",
+		},
+		{
+			"invalid - wrong bech32 prefix",
+			"cosmos1qperwt9wrnkg5k9e5gzfgjppzpqur82k7gqd8n",
+			sdkmath.NewInt(100),
+			"invalid Bech32 prefix; expected kava, got cosmos",
+		},
+		{
+			"invalid - negative amount",
+			"kava1gpxd677pp8zr97xvy3pmgk70a9vcpagsakv0tx",
+			sdkmath.NewInt(-100),
+			"negative amount: -100",
+		},
+		{
+			"invalid - max amount + 1",
+			"kava1gpxd677pp8zr97xvy3pmgk70a9vcpagsakv0tx",
+			types.MAX_FRACTIONAL_AMOUNT.AddRaw(1),
+			"amount exceeds max of 999999999999: 1000000000000",
```

Comment:
It seems like `amount 1000000000000 exceeds max of of 999999999999` would read better and I believe it's best to reserve `:` for cases of wrapped errors